---
title: <center> <h1>RNA-seq analysis of samples expressing mutant TRIM28 gene (TRIM28 6KR)</h1> </center>
output: html_notebook
---

\ 

\ 

```{r load_packages, include=FALSE}
library(BiocManager)
library(knitr)
library(tximport)
library(DESeq2)
library(pheatmap)
library(ggplot2)
library(biomaRt)
library(plotly)
library(RColorBrewer)
library(clusterProfiler)
library(org.Hs.eg.db)
library(pathview)
library(knitr)
library(kableExtra)
library(tidyverse)
```


```{r load_data, include=FALSE}

sample_table = read_csv("sample_table.csv", col_names = TRUE) %>% as.data.frame()

sample_table$treatment <- factor(rep(c("wt_mock", "wt_infected", "6KR_mock", "6KR_infected"), each = 3), levels= c("wt_mock", "wt_infected", "6KR_mock", "6KR_infected"))

sample_files=list.files("C:/Ancsi/Bioinformatics Support Unit/project_virus_2020/salmon_output", full.names = TRUE, recursive = TRUE, pattern = "quant.sf")

names(sample_files) = pull(sample_table, sample_name)

gene_map <- read_csv("../project_2020_gene_map.csv", col_names = c("enst_id", "ensg_id"))

```



```{r 6KR_data_analysis, include=FALSE}

sample_table_6KR = sample_table[7:12, ]

sample_files_6KR = sample_files[7:12]

txi_6KR = tximport(files = sample_files_6KR,
                  type = "salmon",
                  tx2gene = gene_map,
                  ignoreTxVersion = TRUE)

dds_6KR = DESeqDataSetFromTximport(txi = txi_6KR,
                                  colData = sample_table_6KR,
                                  design = ~ treatment)

dds_6KR = DESeq(dds_6KR)

vst_6KR = varianceStabilizingTransformation(dds_6KR)

dist_6KR = as.matrix(dist(t(assay(vst_6KR))))

```


The following table shows the metadata of the experiment, including the samples that express mutant TRIM28 genes.

```{r 6KR_meta, echo=FALSE}
kable(sample_table_6KR) %>% kable_styling(bootstrap_options = c("bordered", "hover"), full_width = F, position = "left")
```

\ 

The heatmap below shows the distances across the samples, which is greatly similar to the first half of the dataset. The mock infected samples are similar to each other and significantly different from the influenza infected samples. The influenza infected samples with the mutant TRIM28 are more similar to each other than the samples expressing the wild type TRIM28.

\ 

```{r 6KR_sample_dist, echo=FALSE}

pheatmap(dist_6KR)

```

\ 

Similarly to the first analysis, the presence of inluenza A is responsible for the great majority of the variance in the dataset, in this case 85% of the variance is explained by the first principal component.

```{r 6KR_plotpca, echo=FALSE}
plotPCA(vst_6KR, intgroup = "treatment")
```


```{r 6KR_result, include=FALSE}

result_dds_6KR = results(dds_6KR, contrast = c("treatment", "6KR_infected","6KR_mock"))

```


The initial result of the differential expression analysis is also highly similar to the first analysis. The number of genes with nonzero read count, low count and outliers are approximately the same resulting in about the same amount of differentially expressed genes based on the default treshold.

```{r 6KR_result_summary, echo=FALSE}
summary(result_dds_6KR)
```


As in the first part of the dataset I introduced the same tresholds for the adjusted p-value and log2 fold change to further filter the initially obtained differentially expressed genes. (adjusted p-value < 0.01 and |log2 fold change| > 2)\ 
Based on this approach there are 2810 differentially expressed genes in this dataset.


```{r 6KR_annotation, include=FALSE}
# adding annotation to the dataset
result_dds_6KR_df = as.data.frame(result_dds_6KR)

dds_6KR_filter1 = result_dds_6KR_df[ complete.cases(result_dds_6KR_df), ]


dds_6KR_filter1 = rownames_to_column(dds_6KR_filter1, var = 'ensg_id')


ensembl100 = useEnsembl(biomart = 'ensembl', dataset = 'hsapiens_gene_ensembl')

annotation_6KR = getBM(attributes = c('ensembl_gene_id',
                                  'entrezgene_id',
                                  'external_gene_name',
                                  'chromosome_name',
                                  'description'),
                  filters = 'ensembl_gene_id',
                  values = dds_6KR_filter1$ensg_id,
                  mart = ensembl100)

annotated_dds_6KR_filter1 = left_join(dds_6KR_filter1, annotation_6KR, by = c('ensg_id' = 'ensembl_gene_id') )
```



```{r 6KR_volcano_script, include= FALSE}

# adding a new column with logical values deciding whether a gene is differentially expressed or not
annotated_dds_6KR_filter1$diff_genes = annotated_dds_6KR_filter1$padj < 0.01 & abs(annotated_dds_6KR_filter1$log2FoldChange) > 2

# setting the shape of data points to triangles for the genes which -log10(adjusted p value) > 50
annotated_dds_6KR_filter1$shape = ifelse((-log10(annotated_dds_6KR_filter1$padj)>50), "triangle", "circle")
annotated_dds_6KR_filter1$limited_padj = -log10(annotated_dds_6KR_filter1$padj)
annotated_dds_6KR_filter1$limited_padj[annotated_dds_6KR_filter1$limited_padj>50] = 50

volcano_plot_6KR = ggplot(annotated_dds_6KR_filter1, aes(x = log2FoldChange,
                                                         y = limited_padj, 
                                                         colour = diff_genes, shape = shape,
                                                         name = external_gene_name)) +
  scale_color_manual(values = c("black", "red")) +
  geom_point(size = 1.5, alpha = 0.2) +
  geom_vline(xintercept = 2, colour = 'grey', linetype = 2) +
  geom_vline(xintercept = -2, colour = 'grey', linetype = 2) +
  geom_hline(yintercept = -log10(0.01), colour = 'grey', linetype = 2) +
  xlim(-10, 10) +
  ylim(0, 50) +
  theme_bw() +
  labs(title = "Volcano plot of mutant TRIM28 samples",
       x = "Log2FoldChange of gene expression",
       y = "-log10 of the adjusted p-value") +
  theme(legend.position = 'none',
        axis.title = element_text(face="bold", size=12, vjust = 0.5),
        plot.title = element_text(hjust = 0.5 ,face="bold", size=14), 
        plot.subtitle = element_text(hjust = 0.5, size=12))

```

\ 

The volcano plot below shows the same features as the volcano plot of the first part of the dataset. The majority of the differentially expressed genes are upregulated and the log2 fold change ranges from 2 to 10. 

```{r 6KR_volcano_plot, echo=FALSE, warning = FALSE}
volcano_plot_6KR
```




```{r 2_6KR_filtering, include=FALSE}

# adding filters to the dataset
dds_6KR_filter2 = dds_6KR_filter1[dds_6KR_filter1$padj < 0.01, ]

dds_6KR_filter3 = dds_6KR_filter2[ abs(dds_6KR_filter2$log2FoldChange) > 2, ]

annotated_dds_6KR_filter3 = left_join(dds_6KR_filter3, annotation_6KR, by = c("ensg_id" = "ensembl_gene_id"))
```



```{r 6KR_top_25_genes, include=FALSE}

# creating a heatmap of the top 25 differentially expressed genes
top_25_6KR = arrange(annotated_dds_6KR_filter3, desc(abs(log2FoldChange)))[1:25, ]

genes_25_6KR = top_25_6KR$ensg_id

names(genes_25_6KR) = top_25_6KR$external_gene_name

top_25_vst_6KR = assay(vst_6KR)[genes_25_6KR, ]

rownames(top_25_vst_6KR) = names(genes_25_6KR)

```



```{r 6KR_colours, include=FALSE}
colors = colorRampPalette(brewer.pal(7, "Blues"))(100)
```

\ 

The heatmap below shows the top 25 differentially expressed genes across the 6 samples.

\ 

```{r 6KR_top_25_heatmap, echo=FALSE}

pheatmap(top_25_vst_6KR, fontsize = 7, color = colors, cutree_cols = 2)

```



```{r 6KR_5, include=FALSE}
# extracting the gene count of the top 5 genes
count_5_6KR = as.data.frame(counts(dds_6KR))

count_5_6KR = rownames_to_column(count_5_6KR, var = "ensg_id")

count_5_6KR =  count_5_6KR[count_5_6KR$ensg_id  %in% genes_25_6KR, ][1:5, ]

count_5_6KR = left_join(count_5_6KR, annotation_6KR[ , c('ensembl_gene_id','external_gene_name')], by = c("ensg_id" = "ensembl_gene_id"))

count_5_6KR = column_to_rownames(count_5_6KR,'external_gene_name' )

count_5_6KR$ensg_id = NULL

# extracting the transformed gene count of the top 5 genes
assay_5_6KR = as.data.frame(assay(vst_6KR))

assay_5_6KR = rownames_to_column(assay_5_6KR, var = "ensg_id")

assay_5_6KR =  assay_5_6KR[assay_5_6KR$ensg_id  %in% genes_25_6KR, ][1:5, ]

assay_5_6KR = left_join(assay_5_6KR, annotation_6KR[ , c('ensembl_gene_id','external_gene_name')], by = c("ensg_id" = "ensembl_gene_id"))

assay_5_6KR = column_to_rownames(assay_5_6KR,'external_gene_name' )

assay_5_6KR$ensg_id = NULL
```

\ 

The tables below shows the difference between the original gene counts and the counts after the variance stabilizing transformation in the top five genes of the heatmap.

Original gene counts:

```{r table_count, echo=FALSE}
kable(count_5_6KR) %>% kable_styling(bootstrap_options = c("bordered", "hover"), full_width = F, position = "left")
```


Transformed gene counts, which were used in the heatmap above:

```{r table_assay, echo=FALSE}
kable(assay_5_6KR) %>% kable_styling(bootstrap_options = c("bordered", "hover"), full_width = F, position = "left")
```




```{r 6KR_gene_set_enrichment, include=FALSE}

# testing for gene set enrichment
entrez_gene_6KR = as.character(annotated_dds_6KR_filter3$entrezgene_id)

entrez_uni_6KR = as.character(annotated_dds_6KR_filter1$entrezgene_id)


ego_6KR = enrichGO(gene = entrez_gene_6KR,
                  OrgDb = org.Hs.eg.db,
                  ont = "BP",
                  universe = entrez_uni_6KR,
                  readable = TRUE)


```


\ 

The following plot shows the top biological processes that are enriched in the differentially expressed genes. Just like in the first half of the dataset the most enriched biological process is "leukocyte differentiation". Beside that only the "positive regulation of leukocyte migration" biological process is present in both datasets in the 10 most enriched gene sets, which indicates significantly different response in the two sample types.

```{r 6KR_dotplot_enrichment, echo=FALSE}
dotplot(ego_6KR)

```



```{r 6KR_kegg, include=FALSE}
# testing for enriched KEGG pathways
kegg_pathway_6KR = enrichKEGG(gene = entrez_gene_6KR,
                             universe = entrez_uni_6KR)

kegg_6KR_df = as.data.frame(kegg_pathway_6KR)

kegg_6KR_main = kegg_6KR_df[ , c("Description", "GeneRatio", "p.adjust")]

rownames(kegg_6KR_main) = NULL


```

\ 

The following table shows the list of over-representated KEGG pathways in the differentially expressed genes.
There are more than twice as many over-represented  pathways in this dataset than in the first one and all over-represented  pathways in the wild type TRIM28 samples are included in the mutant TRIM28 samples.

```{r kegg_table_6KR, echo=FALSE}
kable(kegg_6KR_main) %>% kable_styling(bootstrap_options = c("bordered", "hover"), full_width = F, position = "left")
```

\ 
